File: PhysicMasterApp/app/src/main/java/com/physicmaster/widget/TimerButton.java
package com.physicmaster.widget;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.widget.Button;


import com.physicmaster.R;

import java.util.Timer;
import java.util.TimerTask;


public class TimerButton extends Button {

    private Timer mTimer;
    private int countNumDecrease;
    private int countNumIncrease;
    private boolean isRunning = false;

    /**
     * 定义TimerButton 计时以及停止的时候的字体颜色和背景色
     */
    private int timingBackground, stopBackground, timingTextColor, stopTextColor;

    public boolean isRunning() {
        return 0 != countNumIncrease;

    }

    public TimerButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(attrs);
    }

    public TimerButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs);
    }

    public TimerButton(Context context) {
        super(context);

    }

    private void init(AttributeSet attrs) {

        if (attrs != null) {
            TypedArray typedArrayed = getContext().obtainStyledAttributes(attrs, R.styleable
                    .TimerButton);
            countNumDecrease = typedArrayed.getInt(R.styleable.TimerButton_timecount, 60);
            countNumIncrease = 0;
            typedArrayed.recycle();
        }

    }

    public void onDestory() {
        clearStatus();
    }

    public void onTimerStart() {
        if (isEnabled() && !isRunning) {
            isRunning = true;
            setEnabled(false);
            setBackgroundResource(timingBackground);
            setTextColor(getResources().getColor(timingTextColor));
            mTimer = new Timer();
            mTimer.schedule(new TimerTask() {
                @Override
                public void run() {
                    countNumIncrease++;
                    countNumDecrease--;
                    new Handler(Looper.getMainLooper()).post(new Runnable() {
                        @Override
                        public void run() {
                            if (countNumIncrease == 60) {
                                clearStatus();
                            } else {
                                setText(String.format(getContext().getString(R.string
                                        .transfer_dxyamcode_dy), countNumDecrease));
                            }
                        }

                    });
                }

            }, 10, 1000);
        }

    }

    public void clearStatus() {
        isRunning = false;
        if (mTimer != null) {
            mTimer.cancel();
        }
        setEnabled(true);
        mTimer = null;
        countNumIncrease = 0;
        countNumDecrease = 60;
        setText(getContext().getString(R.string.transfer_dxyamcode));
        // setBackgroundResource(R.drawable.timer_btn_background);
        setBackgroundResource(stopBackground);
        setTextColor(getResources().getColor(stopTextColor));

    }

    /**
     * 设置TimerButton的风格
     */
    public void setButtonStyle(int timingTextColor, int timingBackground, int stopTextColor, int
            stopBackground) {
        this.timingTextColor = timingTextColor;
        this.timingBackground = timingBackground;
        this.stopTextColor = stopTextColor;
        this.stopBackground = stopBackground;
    }

    /**
     * 回收资源
     */
    public void recycle() {
        if (mTimer != null) {
            mTimer.cancel();
        }
    }
}

File: Eyepetizer-in-Kotlin-master/app/src/main/java/com/tt/lvruheng/eyepetizer/widget/TimerButton.kt
package com.tt.lvruheng.eyepetizer.widget

import android.content.Context
import android.os.CountDownTimer
import android.util.AttributeSet
import android.widget.Button
import com.tt.lvruheng.eyepetizer.R

/**
 * TODO: document your custom view class.
 */
class TimerButton : Button {
    private var countDownTimer: CountDownTimer? = null

    constructor(context: Context) : super(context) {
        init(null, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        // Load attributes
        val a = context.obtainStyledAttributes(
                attrs, R.styleable.TimerButton, defStyle, 0)
        a.recycle()
    }

    /**
     * start timing
     */
    fun startTimer(period: Long) {
        countDownTimer = object : CountDownTimer(period, 1 * 1000) {
            override fun onTick(millisUntilFinished: Long) {
                text = "${millisUntilFinished / 1000}秒"
            }

            override fun onFinish() {
                text = "获取验证码"
                isEnabled = true
            }
        }
        isEnabled = false
        countDownTimer!!.start()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        if (countDownTimer != null) {
            countDownTimer!!.cancel()
        }
    }
}
